### ISSUE TITLE
Performance issue with ZeroOrMorePath

### ISSUE THREAD

**Issue by joc-proxem** [dotnetrdf/dotnetrdf#363]
The following code is slow
 ```     
            var graph = new Graph();
            var ttlParser = new TurtleParser();
            ttlParser.Load(graph, @"fibo-vD.ttl");  // https://spec.edmcouncil.org/fibo/vocabulary/master/latest/fibo-vD.ttl
            var dataset = new InMemoryDataset(graph);
            var queryProcessor = new LeviathanQueryProcessor(dataset);
            var queryParser = new SparqlQueryParser();
            var query = queryParser.ParseFromString(@"
SELECT ?concept WHERE {
    ?scheme <http://www.w3.org/2004/02/skos/core#hasTopConcept> ?topConcept.
    ?topConcept ^<http://www.w3.org/2004/02/skos/core#broader>* ?concept.
}");
            queryProcessor.ProcessQuery(query);
```

Much of the time is spent in the following loop (in ZeroOrMorePath.cs)
```
                foreach (ISet s in results.Sets)
                {
                    if (!context.OutputMultiset.Sets.Contains(s))
                    {
                        context.OutputMultiset.Add(s.Copy());
                    }
                }
```
because context.OutputMultiset.Sets is a dynamic IEnumerable that is computed again and again inside the loop

Pre-computing the sets outside the loop yields much faster results : 
```
                var sets = new HashSet<ISet>(context.OutputMultiset.Sets);  // pre-compute here
                foreach (ISet s in results.Sets)
                {
                    //if (!context.OutputMultiset.Sets.Contains(s))
                    if (!sets.Contains(s))  // much faster
                    {
                        sets.Add(s); 
                        context.OutputMultiset.Add(s.Copy());
                    }
                }
```

**Comment by kal** [comment]
Thanks for the report!

I found that the simple hashing approach didn't quite work because the definition of hashcode and equals for sets don't quite match internally. Even so, the work around for this that allows us to make use of the hashing is still much faster for all but the smallest of cases so I think it is worth doing.

**Comment by kal** [comment]
Fixed in release 2.6.2
